### PR TITLE
[jaeger-operator] fix jaeger-operator chart bug

### PR DESCRIPTION
**Pull request by zackzhangkai**
This is a bug when using chart to deploy jaeger-operator.

The selector of service is not match the label of pods.

And after delete the jaeger-operator-metrics service and jaeger-operator pod, only the pod can recreate but the service cannot.

### CODE REVIEW

**Comment by zackzhangkai** [comment]
refered issure: https://github.com/jaegertracing/jaeger-operator/issues/1076

**Comment by naseemkullah** [comment]
cc @cpanato @batazor

**Comment by cpanato** [comment]
@zackzhangkai please bump the chart version

**Comment by zackzhangkai** [comment]
jaeger-operator-1.17.1 and jaeger-operator-1.18.0  have the problem.

**Comment by naseemkullah** [comment]
I don't use jaeger operator but I am quite familiar with helm charts, my two cents is to address the problem here: https://github.com/jaegertracing/helm-charts/blob/fa4dac8d7705fad8a02733aa43019e14dab0b547/charts/jaeger-operator/templates/_helpers.tpl#L45-L48

**Comment by naseemkullah** [comment]
I don't see a reason to stray away from what `helm create` does:

https://github.com/helm/helm/blob/561cc4280830af4a40f2e60c5f17fb3ee69b4e39/pkg/chartutil/create.go#L421-L439

**Comment by naseemkullah** [comment]
Maybe the fix could be addressed in https://github.com/jaegertracing/helm-charts/blob/fa4dac8d7705fad8a02733aa43019e14dab0b547/charts/jaeger-operator/templates/_helpers.tpl#L45-L48 by doing what `helm create` does: 

https://github.com/helm/helm/blob/561cc4280830af4a40f2e60c5f17fb3ee69b4e39/pkg/chartutil/create.go#L421-L439

**Comment by zackzhangkai** [comment]
already fix it as your suggesstion

**Comment by naseemkullah** [comment]
> already fix it as your suggesstion

I don't think that will work. The selector labels are  `app.kubernetes.io/name` and
    `app.kubernetes.io/instance` so the common labels should include at least that, I would suggest to please download latest version of `helm` cli and run `helm create foo` and have a look at the generated helpers, service and deployment templates, then do the same thing here. 😄

**Comment by naseemkullah** [comment]
i.e.

```yaml
{{/*
Common labels
*/}}
{{- define "jaeger-operator.labels" -}}
helm.sh/chart: {{ include "jaeger-operator.chart" . }}
{{ include "jaeger-operator.selectorLabels" . }}
{{- if .Chart.AppVersion }}
app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
{{- end }}
app.kubernetes.io/managed-by: {{ .Release.Service }}
{{- end }}

{{/*
Selector labels
*/}}
{{- define "jaeger-operator.selectorLabels" -}}
app.kubernetes.io/name: {{ include "jaeger-operator.name" . }}
app.kubernetes.io/instance: {{ .Release.Name }}
{{- end }}
```

**Comment by zackzhangkai** [comment]
The service should not be created by this chart  as this [issue](https://github.com/jaegertracing/helm-charts/issues/126), so I remove it.

**Comment by naseemkullah** [comment]
So then `app.kubernetes.io/name` does not need to change, right? Just remove the service template as you've done, but leave the label as is.

File: charts/jaeger-operator/templates/NOTES.txt [hunk 1]
```diff
@@ -2,7 +2,7 @@ jaeger-operator is installed.
 
 
 Check the jaeger-operator logs
-  export POD=$(kubectl get pods-l app.kubernetes.io/instance={{ .Release.Name }} -lapp.kubernetes.io/name=jaeger-operator --namespace {{ .Release.Namespace }} --output name)
+  export POD=$(kubectl get pods-l app.kubernetes.io/instance={{ .Release.Name }} -lname=jaeger-operator --namespace {{ .Release.Namespace }} --output name)
```
Comment (naseemkullah) [inline charts/jaeger-operator/templates/NOTES.txt:5]: Please revert.

File: charts/jaeger-operator/templates/_helpers.tpl [hunk 2]
```diff
@@ -44,5 +44,5 @@ Create chart name and version as used by the chart label.
 
 {{/* Generate basic labels */}}
 {{- define "jaeger-operator.labels" }}
-app.kubernetes.io/name: {{ include "jaeger-operator.name" . }}
```
Comment (naseemkullah) [inline charts/jaeger-operator/templates/_helpers.tpl:47]: Please revert.

**Comment by zackzhangkai** [comment]
This is actually what is the problem. The selector of the service `jaeger-operator-metrics` created by itself , is always `name=jaeger-operator`. 

just as follows :
```
 # k get po jaeger-operator-5dbf4647d5-tkvhx --show-labels
NAME                               READY   STATUS    RESTARTS   AGE   LABELS
jaeger-operator-5dbf4647d5-tkvhx   1/1     Running   0          17s   app.kubernetes.io/name=jaeger-operator,pod-template-hash=5dbf4647d5

# k describe svc jaeger-operator-metrics
Name:              jaeger-operator-metrics
Namespace:         istio-system
Labels:            name=jaeger-operator
Annotations:       <none>
Selector:          name=jaeger-operator
Type:              ClusterIP
IP:                10.233.30.93
Port:              http-metrics  8383/TCP
TargetPort:        8383/TCP
Endpoints:         <none>
Port:              cr-metrics  8686/TCP
TargetPort:        8686/TCP
Endpoints:         <none>
Session Affinity:  None
```

**Comment by naseemkullah** [comment]
> This is actually what is the problem. The selector of the service `jaeger-operator-metrics` created by itself , is always `name=jaeger-operator`.
> 
> just as follows :
> 
> ```
>  # k get po jaeger-operator-5dbf4647d5-tkvhx --show-labels
> NAME                               READY   STATUS    RESTARTS   AGE   LABELS
> jaeger-operator-5dbf4647d5-tkvhx   1/1     Running   0          17s   app.kubernetes.io/name=jaeger-operator,pod-template-hash=5dbf4647d5
> 
> # k describe svc jaeger-operator-metrics
> Name:              jaeger-operator-metrics
> Namespace:         istio-system
> Labels:            name=jaeger-operator
> Annotations:       <none>
> Selector:          name=jaeger-operator
> Type:              ClusterIP
> IP:                10.233.30.93
> Port:              http-metrics  8383/TCP
> TargetPort:        8383/TCP
> Endpoints:         <none>
> Port:              cr-metrics  8686/TCP
> TargetPort:        8686/TCP
> Endpoints:         <none>
> Session Affinity:  None
> ```
> 
> So it is necessary to change the labels of the pods.

Oh! I see ok. I clearly do not know enough about jaeger operator (today, my org uses Jaeger directly via https://github.com/jaegertracing/helm-charts/tree/master/charts/jaeger). I will let @cpanato and @batazor review this one just like all other operator PRs.

**Comment by zackzhangkai** [comment]
Hi, what's going on?

**Comment by naseemkullah** [comment]
@zackzhangkai the PR is awaiting review from jaeger operator chart maintainers.

**Comment by appanycd** [comment]
@batazor @cpanato Can you please review this fix?

**Comment by zackzhangkai** [comment]
Hi, how about this? It has been for a bit long time .  What's going on?  Could you please verify this ? Because our system also has this problem, and we want to be the same with the community . Thank you.

File: charts/jaeger-operator/templates/service.yaml [hunk 3]
```diff
@@ -1,16 +0,0 @@
-apiVersion: v1
```
[the remark this answers is not in the document]
Comment (zackzhangkai) [reply charts/jaeger-operator/templates/service.yaml:1]: this service is created by itself, whenever the backend pod is created, should not install this service mannually.

**Pull request closed without merging**